### PR TITLE
add of auto,diagonal and random feature conceptors

### PR DESCRIPTION
**Pull request by guillaumepourcel**

### CODE REVIEW

File: src/utils/config.py [hunk 1]
```diff
@@ -73,9 +106,19 @@ class TrainingConfig(NamedTuple):
         e.g. regularizer
     """
     washout: int
-    compute_conceptor: Callable
-    compute_loading: Callable
-    init_states: Callable
+    aperture: float
+    wout_regularizer: float
+    #compute_conceptor: Callable
```
Comment (stevenabreu7) [inline src/utils/config.py]: Commenting these out will break my code because I use these fields to determine which function I call to compute the conceptor/loading/etc. 
I will check how you handle these function calls in your pipelines to see how we can merge this properly

File: src/utils/helper.py [hunk 2]
```diff
@@ -1,14 +1,60 @@
 import numpy as np
 
 
-def n_periodic(n, T, key, minval=-1., maxval=1.):
+def n_periodic(n, T, key, minval=-1., maxval=1., snr=0.):
     arr = key.uniform(low=minval, high=maxval, size=(n,))
     arr -= arr.min()
     arr /= arr.max()
-    arr = arr * 2. - 1.
+    arr = (arr * 2. - 1.) * 0.9
```
Comment (stevenabreu7) [inline src/utils/helper.py]: Why multiply it with 0.9?

File: src/esn.py [hunk 3]
```diff
@@ -209,3 +425,209 @@ def update_weights(self, xt: Array, ut: Array, yt_hat: Array,
         :param optimizer: optimizer object, e.g. linear regression.
         """
         self.w_out = self._ridge_regression(xt, ut, yt_hat, alpha=alpha)
+
+
+# Auto-conceptor parts: to factorize (and combined with RCF) after debugged
```
Comment (stevenabreu7) [inline src/esn.py:418]: By now this is debugged, so we can factorize and combine it with RFC right?

File: src/esn.py [hunk 4]
```diff
@@ -29,16 +34,17 @@ def __init__(self, config: ESNConfig, prng: Optional[Any]) -> None:
...
+        self.random_feature = config.random_feature
+        self.random_feature_size = config.random_feature_size
 
         # PRNG key
-        if prng is None:
-            self.prng = np.random.default_rng()
-        else:
-            assert isinstance(prng, np.random.Generator)
-            self.prng = prng
+        assert isinstance(prng, np.random.Generator)
```
Comment (stevenabreu7) [inline src/esn.py:43]: this will raise an error if we move back to jax again. I guess for now we don't care about it, so it's fine

File: src/esn.py [hunk 5]
```diff
@@ -29,16 +34,17 @@ def __init__(self, config: ESNConfig, prng: Optional[Any]) -> None:
...
+        self.prng = prng
 
         # shortcut
-        K, N, L = self.get_sizes()
+        K, N, N_rfc, L = self.get_sizes()
```
Comment (stevenabreu7) [inline src/esn.py]: Just for consistency, we could call this new size **F**?

File: src/utils/config.py [hunk 6]
```diff
@@ -40,21 +61,33 @@ class ESNConfig(NamedTuple):
...
+    init_weights_f_density: float = 1.0
     init_weights_in_density: float = 1.0
     # arguments to init functions
     init_weights__args: dict = {}
+    init_weights_g__args: dict = {}
+    init_weights_f__args: dict = {}
     init_weights_b__args: dict = {}
     init_weights_in__args: dict = {}
+    # architecture parameter
+    architecture: bool = False
```
Comment (stevenabreu7) [inline src/utils/config.py]: variable naming: since these are booleans, it would be better/easier to name them something like `is_top_layer`, `is_random_feature`, etc.

File: src/esn.py [hunk 7]
```diff
@@ -132,19 +179,181 @@ def _forward(self, ut: Array, x_init: Optional[Array] = None,
...
         yt = np.concatenate(yt, axis=1).T
+        
         if collect_states:
             xt = np.concatenate(xt, axis=1).T
             return xt, yt
         else:
             return yt
 
-    def harvest_states(self, ut: Array, x_init: Optional[Callable] = None,
+    def _forwardRFC(self, ut: Array, x_init: Optional[Array] = None,
```
Comment (stevenabreu7) [inline src/esn.py:183]: perhaps it's better to have the type of conceptor be a parameter in the trainingConfiguration (or maybe even esnConfiguration). I would think that we usually know what kind of conceptors we want to use before setting up the ESN in the first place - and then the type of conceptor won't change anymore - so we could just put it into the esnConfig, what do you think?

File: src/esn.py [hunk 8]
```diff
@@ -96,6 +128,21 @@ def normalize_spectral_radius(self, rho: float = 1.0) -> None:
         current_rho = max(abs(np.linalg.eig(self.w)[0]))
         # scale weight matrix to desired spectral radius
         self.w *= rho / current_rho
+    
+    def normalize_spectral_radius_rfc(self, rho: float = 1.0) -> None:
```
Comment (stevenabreu7) [inline src/esn.py]: this can be part of the existing normalisation function, just add a `is_random_feature` boolean parameter in the function call

File: src/esn.py (hunk 7 above)
Comment (stevenabreu7) [reply src/esn.py:183]: because then we can simply call the normal functions (like `forward` or `harvest_states`) while the internal state of the ESN instance will handle the details

File: src/esn.py [hunk 9]
```diff
@@ -132,19 +179,181 @@ def _forward(self, ut: Array, x_init: Optional[Array] = None,
...
+            
+            self.w = W_loaded.copy()
+        
+            Ci = [
+            compute_conceptor(xt_i[trainingConfig.washout:, :], aperture=trainingConfig.aperture)
+            for xt_i in xt
+            ]
+        return Ci
+    
+    def full_procedure_diag(self, ut, trainingConfig: TrainingConfig):
```
Comment (stevenabreu7) [inline src/esn.py:368]: again, I would either
- merge this into the full_procedure function and do if statements for the diagonal conceptor specific parts
- call a function full_procedure which takes a parameter which decides which to call, e.g. full_procedure_rf or full_procedure_diag

File: src/esn.py [hunk 10]
```diff
@@ -154,9 +363,13 @@ def harvest_states(self, ut: Array, x_init: Optional[Callable] = None,
         :return xt: (T, N)
         :return yt: (T, L)
         """
-        return self._forward(ut, x_init, collect_states=True, C=C)
+        if self.random_feature:
```
Comment (stevenabreu7) [inline src/esn.py]: that's what I meant with the comment above, this is nice

File: src/esn.py [hunk 11]
```diff
@@ -178,9 +394,9 @@ def _ridge_regression(self, xt: Array, ut: Array, yt_hat: Array,
             w_out = np.dot(np.linalg.pinv(S), yt_hat).T
         else:
             # ridge regression
-            R = np.dot(S.T, S) / xt.shape[0]
+            R = np.dot(S.T, S) #/ xt.shape[0]
```
Comment (stevenabreu7) [inline src/esn.py]: remove comments here

File: src/esn.py [hunk 12]
```diff
@@ -209,3 +425,209 @@ def update_weights(self, xt: Array, ut: Array, yt_hat: Array,
...
+                Ct.append(C)
+
+        if collect_states:
+            xt = np.concatenate(xt, axis=1).T
+            yt = np.concatenate(yt, axis=1).T
+            return xt, yt, Ct
+        else:
+            return x, y, C
+
+    def cueing(self, ut: Array, learning_rate: float, aperture: float,
```
Comment (stevenabreu7) [inline src/esn.py]: `cueing` and `forward_cueing` only differ in whether or not to collect the states right? then `forward_cueing` and `harvest_cueing` (or smith like that) would be a better name to understand it more easily
```suggestion
    def harvest_cueing(self, ut: Array, learning_rate: float, aperture: float,
```

File: src/esn.py [hunk 13]
```diff
@@ -209,3 +425,209 @@ def update_weights(self, xt: Array, ut: Array, yt_hat: Array,
...
+        :param learning_rate: float, learning rate for incremental adapatation of C
+        :param aperture: float
+        :return x: (N, 1)
+        :return y: (L,1)
+        :return C: (N, N), cue conceptor
+        """
+        return self._auto(ut, learning_rate, aperture, x_init, y_init,
+                          collect_states=False, recall=False)
+
+    def recall(self, ut: Array, learning_rate: float, aperture: float,
```
Comment (stevenabreu7) [inline src/esn.py]: same as above, maybe call this `harvest_recall`?
```suggestion
    def harvest_recall(self, ut: Array, learning_rate: float, aperture: float,
```

**Review by stevenabreu7 — CHANGES_REQUESTED** [review]
For now I only looked at the python files - once we agree on all the changes, we can adapt the notebooks to the new changes and then review the notebooks in this PR (and then finally merge)

File: src/utils/config.py (hunk 1 above)
Comment (guillaumepourcel) [reply src/utils/config.py]: Okay, I did not know you used trainingconfig. Just initializing by default to None ("Callable = None") works for me.

File: src/utils/helper.py (hunk 2 above)
Comment (guillaumepourcel) [reply src/utils/helper.py]: My bad. I tried it bcz Herbert scale everything like that for memory management. But on the rest of the paper there's no specific scaling to 0.9.

File: src/esn.py (hunk 3 above)
Comment (guillaumepourcel) [reply src/esn.py:418]: No, neither RFC nor autoconceptor works well

File: src/esn.py (hunk 5 above)
Comment (guillaumepourcel) [reply src/esn.py]: M is okay? f is used for a matrix in RFC

File: src/esn.py (hunk 9 above)
Comment (guillaumepourcel) [reply src/esn.py:368]: It seems misleading because the procedure of the diagonal is very different from the other one (it's Jong's ugly trick). Maybe just changing the name? After I plan to create a function specific to the different "full_procedure" (but they are more iterative than the classical one) for diagonal conceptors.

File: src/esn.py (hunk 11 above)
Comment (guillaumepourcel) [reply src/esn.py]: Why do you normalize for ridge regression? It is not a correlation matrix

File: src/esn.py (hunk 12 above)
Comment (guillaumepourcel) [reply src/esn.py]: Okay, I changed it.

File: src/esn.py (hunk 13 above)
Comment (guillaumepourcel) [reply src/esn.py]: Okay, I changed it.

File: src/utils/config.py (hunk 1 above)
Comment (stevenabreu7) [reply src/utils/config.py]:
```suggestion
    compute_conceptor: Optional[Callable] = None
```

File: src/esn.py (hunk 5 above)
Comment (stevenabreu7) [reply src/esn.py:47]: 👍

**Comment by stevenabreu7** [comment]
Merged (note: notebooks may not work)
